CARDS-1781: Configure surveys and their frequencies

acrowthe opened the pull request:
Create importer for generic form template
Use importer on PREMs forms (OAIP, OED, CPESIC)

docker-compose test:
- `mvn clean install -Pdocker`
- `python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --cards_project cards4prems --composum --smtps --smtps_test_container --smtps_test_mail_path=<some path>`
- add these to the compose file:
```
    - NIGHTLY_NOTIFICATIONS_SCHEDULE=0 * * * * ? *
    - PATIENT_NOTIFICATION_FROM_ADDRESS=noreply@uhn.ca
    - PATIENT_NOTIFICATION_FROM_NAME=Your Experience @ UHN
```
- `docker-compose build && docker-compose up -d`
- log in as admin
- create a new patient information with a valid email address and consent to receive emails
- create a visit information with all the fields filled in, and a visit date set exactly 7 days ago
- create another visit information with all the fields filled in, with a different clinic, and a visit date set exactly 14 days ago
- create another visit information with all the fields filled in, and a visit date set exactly 10 days ago
- create another visit information with all the fields filled in, status set to `cancelled`, and a visit date set exactly 7 days ago
- wait for one minute
- make sure two emails are sent, for the first 2 visits
- make sure the links in the emails work fine and the patient can fill in the surveys successfully
- make sure the completed visits and forms show up in the right clinic dashboard

acrowthe committed:
CARDS-1781:
Create importer for generic form template
Use importer on PREMs forms (OAIP, OED, CPESIC)

Known issues:
- Conditionals in CPESIC forms are imported but not working properly
- CPESIC version 3 matrixes have some odd behaviour (covering too many questions, missing matrix descriptions)

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/OED.xml:
```diff
@@ -33,28 +33,28 @@
-		<value>1</value>
-		<type>Long</type>
+		<name>paginate</name>
+		<value>True</value>
```
The short forms shouldn't be paginated

acrowthe replied:
They don't have any sections so they don't actually display as paginated. Not sure if that's good enough though

marta- replied:
Right, all good then

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/OED.xml:
```diff
@@ -65,929 +65,1370 @@
 				<name>label</name>
 				<value>No, but I would have liked a reason</value>
 				<type>String</type>
 			</property>
 			<property>
 				<name>value</name>
-				<value>1</value>
+				<value>No, but I would have liked a reason</value>
```
We need to have numerical values for the options for when this data goes for analysis

acrowthe committed:
CARDS-1781: PREMs form import
Update OAIP, OED and CPESIC forms

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/OED.xml:
```diff
@@ -33,9 +33,9 @@
 		<type>Reference</type>
 	</property>
 	<property>
-		<name>maxPerSubject</name>
-		<value>1</value>
-		<type>Long</type>
```
We should actually keep maxpersubject = 1 for all surveys

acrowthe committed:
CARDS-1781: PROMs reimport
- Fix conditionals
- Needs tidying up

acrowthe committed:
CARDS-1781: PREMs form import
- Fully functional conditionals
- Bugs with matrix import causing section crash

- Removed CPESIC version 2 and 3 forms
- Reimported latest version of OED, OAIP and CPESIC

acrowthe committed:
CARDS-1781: PREMs form import
Fix maxPerSubject

acrowthe committed:
CARDS-1781: PREMs form import

acrowthe committed:
CARDS-1781: PREMs form import
- Reimport CPESIC to add multiselect

sashaandjic commented:
problem
https://user-images.githubusercontent.com/23243244/190458960-0cbb1b98-2cd9-4c9e-9501-bf0b6727a808.mp4

marta- commented:
> I created patient with one visit to TGH ER. one survey form was generated. clicking on different clinics in the dashboard ends in error. problem shown in the video.

The bug is in the #1174 branch, it's not related to the surveys.

marta- commented:
> I created patient with one visit to TGH ER. one survey form was generated. clicking on different clinics in the dashboard ends in error. problem shown in the video.

The bug is in the #1174 branch, it's not related to the surveys. Please move the bug report there.

sdumitriu commented:
I think that one part of it is "normal". The way clinic dashboards are set up is:
- they all have the default widgets, visit + one per questionnaire
- they can also load other widgets from an extension point specific to the clinic, but this is optional
- the dashboards seem to load the UI fine, they display the default 2 tables, so the error message is not actually an error
To discuss:
- Is the extension mechanism still useful? It was the only way to populate the clinic dashboard initially, but we later switched to autopopulating it from the clinic mapping / qset
- The logged error should not be a logged error, it is an expected behavior

Second part, I see that the visit is not present in any of the dashboards, but I can't see what the visit information looks like, so I don't know if this is a bug or just an incomplete visit.

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/OAIP.xml:
```diff
@@ -881,29 +866,34 @@
-			<value>Overall...</value>
-			<type>String</type>
+			<name>maxAnswers</name>
+			<value>10</value>
```
```suggestion
			<value>1</value>
```

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/OED.xml:
```diff
@@ -900,29 +907,34 @@
-			<value>Overall...</value>
-			<type>String</type>
+			<name>maxAnswers</name>
+			<value>10</value>
```
```suggestion
			<value>1</value>
```

marta- committed:
CARDS-1781: Configure surveys and their frequencies

Updated maxAnswers for the slider questions in OED and OAIP

marta- committed:
CARDS-1781: Configure surveys and their frequencies

Updated maxAnswers for the slider questions in OED and OAIP

marta- commented:
> I think that one part of it is "normal". The way clinic dashboards are set up is:

I confirm that the console messages are normal.

marta- commented:
> Second part, I see that the visit is not present in any of the dashboards, but I can't see what the visit information looks like, so I don't know if this is a bug or just an incomplete visit.

I can't replicate with data formatted as expected in production.

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/OAIP.xml:
```diff
@@ -652,63 +632,63 @@
 			<property>
 				<name>label</name>
-				<value>Not at all</value>
+				<value>-Not at all</value>
```
Typo
```suggestion
				<value>Not at all</value>
```

marta- committed:
CARDS-1781: Configure surveys and their frequencies

Typo fix

acrowthe committed:
CARDS-1781: PREMs form import

Fix type

marta- committed:
CARDS-1781: Configure surveys and their frequencies
Adjust frequencies for OAIP and ED surveys

sashaandjic commented:
QUESTION: Why do we generate 3 tokens for each visit? I created only one visit, and for that visit the mail was sent with only one token that is highlighted in the list shown below. The other 2 are created for the same subject.
![image](https://user-images.githubusercontent.com/23243244/191932996-e0d8687a-866b-4c00-a2e1-842421193d1c.png)

sashaandjic commented:
Also once I submitted my surveys I should not be seeing this:
![image](https://user-images.githubusercontent.com/23243244/191934287-a075ce24-df70-42f1-9f59-2a6f6a07af7a.png)

sashaandjic commented:
Visiting link provided in mail gives me this:
![image](https://user-images.githubusercontent.com/23243244/191934636-e8846177-b9c4-48ef-a049-4c61300dee0b.png)

however once I remove `https://` all is good.

sashaandjic commented:
Also, I am not receiving any real mails. I am just reading what is written in the file stored in `<some path>` that I specified at the start.

marta- commented:
> Visiting link provided in mail gives me this:

> however once I remove `https://` from url all is good.

This is normal because your locan instance isn't running on https. It won't happen in production which will be set up properly for secure connections.

marta- commented:
> Also once I submitted my surveys I should not be seeing this

I'm looking into it.

sdumitriu commented:
> Why do we generate 3 tokens for each visit?
Because the testing instructions are set up to generate new emails every 60 seconds. Every email generates a token. If you wait 20 minutes, you'll get 20 tokens for each visit. This is "normal" and will not happen in production, where each visit gets one (or 2) emails.

sdumitriu commented:
> Also, I am not receiving any real mails. I am just reading what is written in the file stored in <some path> that I specified at the start.

When you say you don't receive real emails, you mean to your actual email account? That's normal, the testing instructions only set up a fake email server that dumps emails to the filesystem. If you can see the emails in that folder, that's all there is to see.

sdumitriu commented:
> Here are some instructions if you want to test emailing: https://github.com/data-team-uhn/cards/pull/1067. @sdumitriu are these still up to date after your recent fixes?

The same instructions are in this PR, which I think Sasha already followed.

sashaandjic commented:
> > Also once I submitted my surveys I should not be seeing this:
> 
> @sashaandjic Is this the only outstanding issue?

yes

marta- commented:
> > > Also once I submitted my surveys I should not be seeing this:
> > 
> > 
> > @sashaandjic Is this the only outstanding issue?
> 
> yes

Addressed in #1174 in commit [4fea46d](https://github.com/data-team-uhn/cards/pull/1174/commits/4fea46d196f97d79884f1f4ed66fe5877d163fad). This branch has been rebased to include the change.
